tests/unit/test_parsers_jats.py

```diff
@@ -96,7 +98,10 @@ def test_field(field_name, records):
     result = getattr(records['jats'], field_name)
     expected = records['expected'][field_name]
 
-    assert result == expected
+    diffs = DeepDiff(result, expected, ignore_order=True)
+    if sys.version_info[0] < 3 and 'type_changes' in diffs:
+        del diffs['type_changes']
```

Comment:
do you really need `DeepDiff` here? if nothing but the string type has changed, there won't be an issue, as
`"foo" == u"foo"` (with ASCII only).

Reply:
the expected output is read from `YAML`. yaml.load() only marks non-ascii as `u''`
the `jats` parser marks everything as `u''`.

for string comparison this doesn't matter, but here we compare nested data structure with lists of dicts

I use DeepDiff to compare records with `ignore_order=True` because there are lists of dicts (e.g. affiliations) which do not compare as equal unless they are in the same order, however in python3 you can't sort a list of dicts easily.

```python3
Python 3.7.2 (default, Mar 21 2019, 10:09:12) 
>>> sorted([{2:3,4:5}, {3:1, 1:2}])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'dict' and 'dict'
>>> 
```
```python2
Python 2.7.15 (default, Oct 15 2018, 15:26:09) 
>>> sorted([{2:3,4:5}, {3:1, 1:2}])
[{1: 2, 3: 1}, {2: 3, 4: 5}]
```

DeepDiff does do a type comparison, and in python2 the types are different

```python2
>>> type('hello')
<type 'str'>
>>> type(u'hello')
<type 'unicode'>
```

while in python3 
```
>>> type('hello')
<class 'str'>
>>> type(u'hello')
<class 'str'>
```

DeepDiff has a `ignore-type` option in current version, but the last version compatible with python2.7 doesn't have that option.

Reply:
@michamos 

after changing this to `assert expected == result` I get occasional test errors on subsequent runs -- seemingly depending on some external state of machine 

`$ pytest -vv tests/unit/test_parsers_jats.py`

it doesn't help to do `rm -rf .pytest_cache/`

```
field_name = 'authors'
records = {'expected': {'abstract': 'The study of non-Abelian Majorana zero modes advances our understanding of the fundamental ...'}]}], ...}, 'file_name': 'PhysRevX.4.021018.xml', 'jats': <hepcrawl.parsers.jats.JatsParser object at 0x7ffb2dc08390>}

    @pytest.mark.parametrize(
        'field_name',
        FIELDS_TO_CHECK
    )
    def test_field(field_name, records):
        result = getattr(records['jats'], field_name)
        expected = records['expected'][field_name]
    
>       assert expected == result
E       AssertionError: assert [{'full_name'...ng, China'}]}] == [{'full_name':...ng, China'}]}]
E         At index 0 diff: {'raw_affiliations': [{'source': 'American Physical Society', 'value': 'Department of Physics, Hong Kong University of Science and Technology, Clear Water Bay, Hong Kong, China'}, {'source': 'American Physical Society', 'value': 'Institute for Advanced Study, Hong Kong University of Science and Technology, Clear Water Bay, Hong Kong, China'}], 'full_name': 'Liu, Xiong-Jun'} != {'full_name': 'Liu, Xiong-Jun', 'raw_affiliations': [{'value': 'Institute for Advanced Study, Hong Kong University of Science and Technology, Clear Water Bay, Hong Kong, China', 'source': 'American Physical Society'}, {'value': 'Department of Physics, Hong Kong University of Science and Technology, Clear Water Bay, Hong Kong, China', 'source': 'American Physical Society'}]}
E         Full diff:
E         [{'full_name': 'Liu, Xiong-Jun',
E         'raw_affiliations': [{'source': 'American Physical Society',
E         +                         'value': 'Institute for Advanced Study, Hong Kong '
E         +                                  'University of Science and Technology, Clear '
E         +                                  'Water Bay, Hong Kong, China'},
E         +                        {'source': 'American Physical Society',
E         'value': 'Department of Physics, Hong Kong University '
E         'of Science and Technology, Clear Water Bay, '
E         -                                  'Hong Kong, China'},
E         +                                  'Hong Kong, China'}]},
E         ?                                                     ++
E         -                        {'source': 'American Physical Society',
E         -                         'value': 'Institute for Advanced Study, Hong Kong '
E         -                                  'University of Science and Technology, Clear '
E         -                                  'Water Bay, Hong Kong, China'}]},
E         {'full_name': 'Wong, Chris L.M.',
E         'raw_affiliations': [{'source': 'American Physical Society',
E         'value': 'Department of Physics, Hong Kong University '
E         'of Science and Technology, Clear Water Bay, '
E         'Hong Kong, China'}]},
E         {'full_name': 'Law, K.T.',
E         'raw_affiliations': [{'source': 'American Physical Society',
E         'value': 'Department of Physics, Hong Kong University '
E         'of Science and Technology, Clear Water Bay, '
E         'Hong Kong, China'}]}]

tests/unit/test_parsers_jats.py:101: AssertionError
... 
====================================== 1 failed, 74 passed, 3 skipped, 2 warnings in 18.99 seconds ======================================
```
 
it is always the author field that errors

```
for x in `seq 1 10`; do pytest tests/unit/test_parsers_jats.py ; done
```
gives  test pass/fail   `+ - + - + + + + - -`  so there is no discernible pattern

trying with cache removal:
```
for x in `seq 1 10; do rm -rf .pytest_cache; pytest tests/unit/test_parsers_jats.py  |  grep seconds; done
```
`+ + + - - - + - - -`

typically the state changes when I do a bit of mouse movement !

T.

Reply:
so I suggest to make clear the `authors` test is special

```
    if field_name == 'authors':
        diffs = DeepDiff(result, expected, ignore_order=True)
        if sys.version_info[0] < 3 and 'type_changes' in diffs:
            del diffs['type_changes']
        assert diffs == {}
    else:
        assert result == expected
```
 and leave it to a future improvement to figure out stable sort order